File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/repositories/impl/ExampleCommandRepository.java
package com.oddly.ddd.infrastructure.repositories.impl;

import com.oddly.ddd.application.mappers.ExampleMapper;
import com.oddly.ddd.domain.models.ExampleModel;
import com.oddly.ddd.infrastructure.persistence.write.ExampleWriteEntity;
import com.oddly.ddd.infrastructure.repositories.IExampleCommandRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.springframework.stereotype.Repository;

import java.util.concurrent.CompletableFuture;

/**
 * Command repository implementation for Example domain.
 * Repository implementations:
 * - Live in /infrastructure/repositories/impl/
 * - Have Repository suffix
 * - Map BMO ↔ WriteEntity internally
 * - Return void or ID from commands
 */
@Repository
public class ExampleCommandRepository implements IExampleCommandRepository {
    @PersistenceContext
    private EntityManager m_entityManager;

    private final ExampleMapper m_mapper;

    public ExampleCommandRepository(ExampleMapper p_mapper) {
        this.m_mapper = p_mapper;
    }

    @Override
    public CompletableFuture<String> saveAsync(ExampleModel p_model) {
        return CompletableFuture.supplyAsync(() -> {
            p_model.validate();
            ExampleWriteEntity entity = m_mapper.toWriteEntity(p_model);
            m_entityManager.persist(entity);
            return entity.getId();
        });
    }

    @Override
    public CompletableFuture<Void> updateAsync(ExampleModel p_model) {
        return CompletableFuture.runAsync(() -> {
            p_model.validate();
            ExampleWriteEntity entity = m_mapper.toWriteEntity(p_model);
            m_entityManager.merge(entity);
        });
    }

    @Override
    public CompletableFuture<Void> deleteAsync(String p_id) {
        return CompletableFuture.runAsync(() -> {
            ExampleWriteEntity entity = m_entityManager.find(ExampleWriteEntity.class, p_id);
            if (entity != null) {
                m_entityManager.remove(entity);
            }
        });
    }

    @Override
    public CompletableFuture<Boolean> existsAsync(String p_id) {
        return CompletableFuture.supplyAsync(() -> {
            Long count = m_entityManager.createQuery(
                "SELECT COUNT(e) FROM ExampleWriteEntity e WHERE e.m_id = :id", Long.class)
                .setParameter("id", p_id)
                .getSingleResult();
            return count > 0;
        });
    }
}

File: infrastructures/ddd/python/src/infrastructure/repositories/impl/example_command_repository.py
"""
Example command repository implementation.
Repository implementations:
- Live in /infrastructure/repositories/impl/
- Implement repository interfaces
- Handle persistence operations
- Map BMO ↔ WriteEntity internally using mapper
"""

from typing import Optional
from application.mappers.example_mapper import ExampleMapper
from domain.models.example_model import ExampleModel
from infrastructure.repositories.i_example_command_repository import IExampleCommandRepository
from infrastructure.persistence.write.example_write_entity import ExampleWriteEntity


class ExampleCommandRepository(IExampleCommandRepository):
    """
    Command repository implementation for Example domain.
    
    This implementation:
    - Uses ExampleMapper for BMO ↔ WriteEntity conversions
    - Handles write operations (create, update, delete)
    - Works with WriteEntity internally
    - Accepts/returns BMO externally
    
    In a real application, this would use a database driver:
    - MongoDB: motor (async) or pymongo
    - PostgreSQL: asyncpg or psycopg2 with SQLAlchemy
    - MySQL: aiomysql with SQLAlchemy
    
    Example with motor (MongoDB):
    ```python
    from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorDatabase
    
    class ExampleCommandRepository(IExampleCommandRepository):
        def __init__(self, p_database: AsyncIOMotorDatabase, p_mapper: ExampleMapper):
            self._collection = p_database["examples"]
            self._mapper = p_mapper
        
        async def save_async(self, p_model: ExampleModel) -> str:
            entity = self._mapper.to_write_entity(p_model)
            result = await self._collection.insert_one({
                "_id": entity.id,
                "name": entity.name,
                "description": entity.description,
                "owner_id": entity.owner_id,
                "is_active": entity.is_active,
                "created_at": entity.created_at,
                "updated_at": entity.updated_at,
                "version": entity.version
            })
            return result.inserted_id
    ```
    """
    
    def __init__(self, p_mapper: ExampleMapper):
        """
        Initialize the command repository.
        
        Args:
            p_mapper: The mapper for BMO ↔ WriteEntity conversions
        """
        self._mapper = p_mapper
        # In real implementation, inject database connection here
        self._in_memory_store: dict[str, ExampleWriteEntity] = {}
    
    async def save_async(self, p_model: ExampleModel) -> str:
        """
        Saves a new example to the database.
        
        Args:
            p_model: The business model to save
            
        Returns:
            The ID of the created entity
        """
        # Map BMO to WriteEntity
        entity = self._mapper.to_write_entity(p_model)
        
        # In real implementation, persist to database
        self._in_memory_store[entity.id] = entity
        
        return entity.id
    
    async def update_async(self, p_model: ExampleModel) -> None:
        """
        Updates an existing example in the database.
        
        Args:
            p_model: The business model to update
        """
        # Map BMO to WriteEntity
        entity = self._mapper.to_write_entity(p_model)
        
        # In real implementation, update in database
        if entity.id in self._in_memory_store:
            self._in_memory_store[entity.id] = entity
        else:
            raise ValueError(f"Example {entity.id} not found")
    
    async def delete_async(self, p_id: str) -> None:
        """
        Deletes an example from the database.
        
        Args:
            p_id: The ID of the example to delete
        """
        # In real implementation, delete from database
        if p_id in self._in_memory_store:
            del self._in_memory_store[p_id]
        else:
            raise ValueError(f"Example {p_id} not found")
    
    async def exists_async(self, p_id: str) -> bool:
        """
        Checks if an example exists.
        
        Args:
            p_id: The ID to check
            
        Returns:
            True if exists, false otherwise
        """
        # In real implementation, check database
        return p_id in self._in_memory_store
    
    async def find_by_id_for_command_async(self, p_id: str) -> ExampleModel:
        """
        Finds an example by ID for command operations.
        
        Args:
            p_id: The ID to find
            
        Returns:
            The business model object
            
        Raises:
            ValueError: If not found
        """
        # In real implementation, query database
        entity = self._in_memory_store.get(p_id)
        
        if entity is None:
            raise ValueError(f"Example {p_id} not found")
        
        # Map WriteEntity to BMO
        return self._mapper.to_model_from_write_entity(entity)
